piControlMain.c

```diff
@@ -781,40 +781,48 @@ static long piControlIoctl(struct file *file, unsigned int prg_nr, unsigned long
 
 	case KB_GET_DEVICE_INFO_LIST:
 		{
-			SDeviceInfo *pDev = (SDeviceInfo *) usr_addr;
+			SDeviceInfo *dev_infos;
+			unsigned int num_devs = RevPiDevice_getDevCnt();
 			bool firmware_update = false;
 			int i;
 
-			for (i = 0; i < RevPiDevice_getDevCnt(); i++) {
-				pDev[i].i8uAddress = RevPiDevice_getDev(i)->i8uAddress;
-				pDev[i].i8uActive = RevPiDevice_getDev(i)->i8uActive;
-				pDev[i].i32uSerialnumber = RevPiDevice_getDev(i)->sId.i32uSerialnumber;
-				pDev[i].i16uModuleType = RevPiDevice_getDev(i)->sId.i16uModulType;
-				pDev[i].i16uHW_Revision = RevPiDevice_getDev(i)->sId.i16uHW_Revision;
-				pDev[i].i16uSW_Major = RevPiDevice_getDev(i)->sId.i16uSW_Major;
-				pDev[i].i16uSW_Minor = RevPiDevice_getDev(i)->sId.i16uSW_Minor;
-				pDev[i].i32uSVN_Revision = RevPiDevice_getDev(i)->sId.i32uSVN_Revision;
-				pDev[i].i16uInputLength = RevPiDevice_getDev(i)->sId.i16uFBS_InputLength;
-				pDev[i].i16uInputOffset = RevPiDevice_getDev(i)->i16uInputOffset;
-				pDev[i].i16uOutputLength = RevPiDevice_getDev(i)->sId.i16uFBS_OutputLength;
-				pDev[i].i16uOutputOffset = RevPiDevice_getDev(i)->i16uOutputOffset;
-				pDev[i].i16uConfigLength = RevPiDevice_getDev(i)->i16uConfigLength;
-				pDev[i].i16uConfigOffset = RevPiDevice_getDev(i)->i16uConfigOffset;
-				pDev[i].i8uModuleState = RevPiDevice_getDev(i)->i8uModuleState;
-
-				if (	pDev[i].i16uModuleType == KUNBUS_FW_DESCR_TYP_PI_DIO_14
-				||	pDev[i].i16uModuleType == KUNBUS_FW_DESCR_TYP_PI_DO_16
-				||	pDev[i].i16uModuleType == KUNBUS_FW_DESCR_TYP_PI_DI_16) {
+			dev_infos = kcalloc(num_devs, sizeof(SDeviceInfo),
+					    GFP_KERNEL);
+			if (!dev_infos) {
+				pr_err("failed to allocate memory for device list\n");
```

Comment:
`kmalloc()` failure already causes a splat in dmesg, so an additional error message is unnecessary. These messages have largely been removed in the mainline kernel by tons of small janitorial patches, see torvalds/linux@b2ae5d4dc7562788fcba1cb51cf002f28de91133 for an example.

Reply:
Agreed, I will remove it

Reply:
You can force-push to your branch and GitHub will automatically update the pull request.

Nice work overall, thanks a lot.

Reply:
Will do, thanks for the review!